pkg/cloud/vsphere/services/govmomi/util.go

```diff
@@ -165,3 +167,76 @@ func getExistingMetadata(ctx *context.MachineContext) (string, error) {
 
 	return string(metadataBuf), nil
 }
+
+// lookupVM returns the VM, an RequeueError because the VM has an in-flight,
+// or an unexpected error.
+func lookupVM(ctx *context.MachineContext) (*object.VirtualMachine, error) {
+
+	// Check to see if there is an in-flight task.
+	if task := getTask(ctx); task == nil {
+		ctx.MachineStatus.TaskRef = ""
+	} else {
+		ctx := context.NewMachineLoggerContext(ctx, task.Reference().Value)
+
+		// Since a task was discovered, let's find out if it indicates a VM is
+		// being, or has been, created/cloned.
+		ctx.Logger.V(4).Info("task found", "state", task.Info.State)
```

Comment:
By the way @sidharthsurana, I had the realization last night (not yet having time to change the code to reflect this realization), that we actually don't care at all about the task. Whether it's queued or running or anything. I suppose we should check to see if it exists, and if not then delete the ref. But regardless, we only care *if* there **is** a task. If so, we requeue. And we requeue until there is no task and then process the resultant state from there.

Thoughts?

Reply:
One case that comes to mind is race between the user issues a delete (while the clone task was running) and the time the actuator got into action to delete. Now, if the task is running then we want to cancel it. if it succeeded in the small duration, then we got to find and delete the VM directly.

_thinking about any other situations as well..._

Reply:
Hi @sidharthsurana,

So task cancellation is the one case, but we don't currently support that. But given this case, at worst the create must complete before the delete then occurs. It's not the worst thing. 

If there's nothing other than that, this would greatly simplify understanding the code. And we could always add the exceptions in as we encounter them.

Reply:
By the way, task cancellation is something CAPI's model would have to support as well. Otherwise I'm pretty sure the controller won't process the next call before the previous one is complete.

To that end, what you're discussing is handling the case of the task being cancelled out from under Kubernetes. It's a fair point, but I don't think we care about that right.

However one issue with my proposal it that I describe removing the task ref only when the task can't be found. That's problem since the task ref may live for a while in the collector. We still need to look at the task state and remove the task ref if the task is completed, success or failure.

Reply:
Hi @sidharthsurana,

I updated the commit. I didn't change much. I just simplified the result of success. We don't need to set the MoRef here. That will happen (or won't) when the FindVM call runs below.